File: wellpt-ptprd/wellpt-app-workflow/src/main/java/com/wellsoft/pt/app/workflow/ext/dms/file/web/WorkflowFileViewerController.java
/*
 * @(#)2018年9月26日 V1.0
 *
 * Copyright 2018 WELL-SOFT, Inc. All rights reserved.
 */
package com.wellsoft.pt.app.workflow.ext.dms.file.web;

import com.wellsoft.context.enums.Separator;
import com.wellsoft.pt.app.workflow.ext.dms.file.viewer.WorkflowFileViewer;
import com.wellsoft.pt.app.workflow.web.WorkV53Controller;
import com.wellsoft.pt.bpm.engine.entity.FlowInstance;
import com.wellsoft.pt.bpm.engine.entity.FlowInstanceParameter;
import com.wellsoft.pt.bpm.engine.entity.TaskInstance;
import com.wellsoft.pt.bpm.engine.service.FlowService;
import com.wellsoft.pt.bpm.engine.service.TaskService;
import com.wellsoft.pt.dms.facade.api.DmsFileServiceFacade;
import com.wellsoft.pt.dms.file.action.FileActions;
import com.wellsoft.pt.security.acl.service.AclConstants;
import com.wellsoft.pt.security.core.userdetails.UserDetails;
import com.wellsoft.pt.security.util.SpringSecurityUtils;
import com.wellsoft.pt.workflow.work.bean.WorkBean;
import com.wellsoft.pt.workflow.work.service.WorkService;
import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.lang.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.util.List;

/**
 * Description: 如何描述该类
 *
 * @author zhulh
 * @version 1.0
 *
 * <pre>
 * 修改记录:
 * 修改后版本	修改人		修改日期			修改内容
 * 2018年9月26日.1	zhulh		2018年9月26日		Create
 * </pre>
 * @date 2018年9月26日
 */
@Controller
@RequestMapping(WorkflowFileViewer.WORKFLOW_FILE_VIEWER_PREFIX)
public class WorkflowFileViewerController extends WorkV53Controller {

    @Autowired
    private WorkService workService;

    @Autowired
    private FlowService flowService;

    @Autowired
    private TaskService taskService;

    @Autowired
    private DmsFileServiceFacade dmsFileServiceFacade;

    /**
     * 查看归档的流程数据
     *
     * @param flowInstUuid
     * @param fileUuid
     * @param folderUuid
     * @param model
     * @return
     */
    @RequestMapping(value = "/view/archive", method = RequestMethod.GET)
    public String viewArchive(
            @RequestParam(value = "flowInstUuid") String flowInstUuid,
            @RequestParam(value = "fileUuid") String fileUuid,
            @RequestParam(value = "folderUuid") String folderUuid,
            @RequestParam(value = "_requestCode", defaultValue = StringUtils.EMPTY, required = false) String requestCode,
            HttpServletRequest request, HttpServletResponse response, Model model,
            final RedirectAttributes redirectAttributes) {
        UserDetails user = SpringSecurityUtils.getCurrentUser();
        String requestCodeParamUri = "&_requestCode=" + requestCode;
        TaskInstance taskInstance = taskService.getLastTaskInstanceByFlowInstUuid(flowInstUuid);
        String taskInstUuid = taskInstance.getUuid();
        // 流程权限判断
        if (taskService.hasViewPermission(user, taskInstUuid)) {
            String redirectUrl = "/workflow/work/v53/view/work?taskInstUuid=" + taskInstUuid + "&flowInstUuid="
                    + flowInstUuid + requestCodeParamUri;
            addEpRedirectAttributes(request, redirectAttributes);
            return redirect(redirectUrl);
        }

        // 文件读取权限判断
        if (!dmsFileServiceFacade.hasFilePermission(fileUuid, FileActions.READ_FILE)) {
            return forward(AclConstants.DEFAULT_ERROR_VIEW);
        }

        WorkBean workBean = workService.getRead(taskInstUuid, flowInstUuid);
        model.addAttribute(workBean);
        return forward(workViewV53);
    }

    /**
     * 查看归档的流程表单数据
     *
     * @param flowInstUuid
     * @param fileUuid
     * @param folderUuid
     * @param model
     * @return
     */
    @RequestMapping(value = "/view/dyform/archive", method = RequestMethod.GET)
    public String viewDyformArchive(
            @RequestParam(value = "dataUuid") String dataUuid,
            @RequestParam(value = "fileUuid") String fileUuid,
            @RequestParam(value = "folderUuid") String folderUuid,
            @RequestParam(value = "_requestCode", defaultValue = StringUtils.EMPTY, required = false) String requestCode,
            HttpServletRequest request, HttpServletResponse response, Model model,
            final RedirectAttributes redirectAttributes) {
        UserDetails user = SpringSecurityUtils.getCurrentUser();
        // 获取规档的流程
        String flowInstUuid = getArchiveFlowInstUuid(fileUuid, folderUuid);
        if (StringUtils.isBlank(flowInstUuid)) {
            return forward(AclConstants.DEFAULT_ERROR_VIEW);
        }

        FlowInstance flowInstance = flowService.getFlowInstance(flowInstUuid);
        if (!StringUtils.equals(flowInstance.getDataUuid(), dataUuid)) {
            return forward(AclConstants.DEFAULT_ERROR_VIEW);
        }

        String requestCodeParamUri = "&_requestCode=" + requestCode;
        TaskInstance taskInstance = taskService.getLastTaskInstanceByFlowInstUuid(flowInstUuid);
        String taskInstUuid = taskInstance.getUuid();
        // 流程权限判断
        if (taskService.hasViewPermission(user, taskInstUuid)) {
            String redirectUrl = "/workflow/work/v53/view/work?taskInstUuid=" + taskInstUuid + "&flowInstUuid="
                    + flowInstUuid + requestCodeParamUri;
            addEpRedirectAttributes(request, redirectAttributes);
            return redirect(redirectUrl);
        }

        // 文件读取权限判断
        if (!dmsFileServiceFacade.hasFilePermission(fileUuid, FileActions.READ_FILE)) {
            return forward(AclConstants.DEFAULT_ERROR_VIEW);
        }

        WorkBean workBean = workService.getRead(taskInstUuid, flowInstUuid);
        model.addAttribute(workBean);
        return forward(workViewV53);
    }

    /**
     * @param folderUuid
     * @return
     */
    private String getArchiveFlowInstUuid(String fileUuid, String folderUuid) {
        FlowInstanceParameter example = new FlowInstanceParameter();
        example.setName(folderUuid + Separator.SLASH.getValue() + fileUuid);
        List<FlowInstanceParameter> parameters = flowService.findFlowInstanceParameter(example);
        if (CollectionUtils.isEmpty(parameters)) {
            return null;
        }
        return parameters.get(0).getValue();
    }

}

File: wellpt-front-end/wellapp-workflow/app/controller/WorkflowFileViewerController.js
'use strict';

const Controller = require('wellapp-framework').Controller;
const constant = require('./constant');

class WorkflowFileViewerController extends Controller {
  async viewArchive() {
    const ctx = this.ctx;
    const { flowInstUuid, fileUuid, folderUuid, _requestCode } = ctx.query;

    const taskInstUuid = await this.getTaskInstUuid(flowInstUuid);

    // 流程权限判断
    const hasTaskViewPermission = await ctx.service.workflowTaskService.hasViewPermission(taskInstUuid);
    if (hasTaskViewPermission) {
      const redirectUrl = `/workflow/work/view/work?taskInstUuid=${taskInstUuid}&flowInstUuid=${flowInstUuid}&_requestCode=${_requestCode}`;
      ctx.redirect(redirectUrl);
      return;
    }

    // 文件读取权限判断
    const hasFilePermission = await this.hasFilePermission(fileUuid);
    if (!hasFilePermission) {
      ctx.redirect('/error?real_status=404');
      return;
    }

    const workBean = await ctx.service.workflowTaskService.getReadWorkData(taskInstUuid, flowInstUuid);
    // const workBean = await ctx.service.jsonDataService._call('workService', 'getRead', [taskInstUuid, flowInstUuid]);
    if (workBean) {
      await this.workView(workBean);
    } else {
      ctx.redirect('/error?real_status=404');
    }
  }

  async viewDyformArchive() {
    const ctx = this.ctx;
    const { dataUuid, fileUuid, folderUuid, _requestCode } = ctx.query;

    const flowInstUuid = await this.getArchiveFlowInstUuid(fileUuid, folderUuid);
    if (!flowInstUuid) {
      ctx.redirect('/error?real_status=404');
      return;
    }

    const taskInstUuid = await this.getTaskInstUuid(flowInstUuid);

    // 流程权限判断
    const hasTaskViewPermission = await ctx.service.workflowTaskService.hasPermission(taskInstUuid, [
      constant.AclPermission.FLAG_READ.mask
    ]);
    if (hasTaskViewPermission) {
      const redirectUrl = `/workflow/work/view/work?taskInstUuid=${taskInstUuid}&flowInstUuid=${flowInstUuid}&_requestCode=${_requestCode}`;
      ctx.redirect(redirectUrl);
      return;
    }

    // 文件读取权限判断
    const hasFilePermission = await this.hasFilePermission(fileUuid);
    if (!hasFilePermission) {
      ctx.redirect('/error?real_status=404');
      return;
    }
    const workBean = await ctx.service.workflowTaskService.getReadWorkData(taskInstUuid, flowInstUuid);
    // const workBean = await ctx.service.jsonDataService._call('workService', 'getRead', [taskInstUuid, flowInstUuid]);
    if (workBean) {
      // 表单归档只能查看表单数据及打印表单
      workBean.buttons = workBean.buttons && workBean.buttons.filter(button => button.id == 'B004096');
      await this.workView(workBean);
    } else {
      ctx.redirect('/error?real_status=404');
    }
  }

  async getTaskInstUuid(flowInstUuid) {
    return await this.ctx.service.jsonDataService._call('taskService', 'getLastTaskInstanceUuidByFlowInstUuid', [flowInstUuid]);
  }

  async hasFilePermission(fileUuid) {
    return await this.ctx.service.jsonDataService._call('dmsFileActionService', 'hasFilePermission', [fileUuid, ['readFile']]);
  }

  async getArchiveFlowInstUuid(fileUuid, folderUuid) {
    const flowInstanceParameter = { name: folderUuid + '/' + fileUuid };

    const flowInstanceParameters = await this.ctx.service.jsonDataService._call('flowService', 'findFlowInstanceParameter', [
      flowInstanceParameter
    ]);

    if (Array.isArray(flowInstanceParameters) && flowInstanceParameters.length) {
      return flowInstanceParameters[0].flowInstUuid;
    }

    return null;
  }

  async workView(workBean) {
    const ctx = this.ctx;
    let SYSTEM_ID = ctx.SYSTEM_ID && ctx.SYSTEM_ID !== 'undefined' ? ctx.SYSTEM_ID : undefined;

    let settings = await ctx.service.workflowTaskService.getFlowSettings(SYSTEM_ID);

    await ctx.render('workflow-work/work_view.js', {
      workBean,
      SYSTEM_ID,
      settings
    });
  }
}

module.exports = WorkflowFileViewerController;
